src/libpmemfile-core/block_array.c

```diff
@@ -0,0 +1,230 @@
+
+	return binfo->idx < binfo->arr->length;
+}
+
+static void
+allocate_new_block_array(struct pmemfile_vinode *vinode)
+{
+	struct pmemfile_inode *inode = D_RW(vinode->inode);
+
+	TOID(struct pmemfile_block_array) new =
+			TX_ZALLOC(struct pmemfile_block_array, FILE_PAGE_SIZE);
+	D_RW(new)->length = (uint32_t)
+			((page_rounddown(pmemobj_alloc_usable_size(new.oid)) -
+			sizeof(struct pmemfile_block_array)) /
+			sizeof(struct pmemfile_block));
+
+	D_RW(new)->next = inode->file_data.blocks.next;
+	TX_SET_DIRECT(&inode->file_data.blocks, next, new);
+	vinode->first_free_block.arr = D_RW(new);
+	vinode->first_free_block.idx = 0;
+}
+
+static struct pmemfile_block *
+acquire_new_entry(struct pmemfile_vinode *vinode)
+{
+	struct block_info *binfo = &vinode->first_free_block;
+
+	return binfo->arr->blocks + binfo->idx++;
+}
+
+struct pmemfile_block *
+allocate_new_block_metadata(struct pmemfile_vinode *vinode)
+{
+	update_binfo(vinode);
+
+	if (!has_free_block_entry(vinode))
+		allocate_new_block_array(vinode);
+
+	return acquire_new_entry(vinode);
+}
+
+static struct pmemfile_block *
+last_used_block(struct pmemfile_vinode *vinode)
+{
+	struct block_info *binfo = &vinode->first_free_block;
+
+	ASSERT(binfo->idx > 0);
+
+	return binfo->arr->blocks + (binfo->idx - 1);
+}
+
+static void
+unlink_block(struct pmemfile_block *block)
+{
+	if (!TOID_IS_NULL(block->prev)) {
+		TX_ADD_FIELD_DIRECT(D_RW(block->prev), next);
+		D_RW(block->prev)->next = block->next;
+	}
+
+	if (!TOID_IS_NULL(block->next)) {
+		TX_ADD_FIELD_DIRECT(D_RW(block->next), prev);
+		D_RW(block->next)->prev = block->prev;
+	}
+}
+
+static void
+relocate_block(struct pmemfile_block *dst, struct pmemfile_block *src)
+{
+	ASSERT(dst != src);
+
+	TX_ADD_DIRECT(dst);
+
+	if (!TOID_IS_NULL(src->prev)) {
+		TX_ADD_FIELD_DIRECT(D_RW(src->prev), next);
+		D_RW(src->prev)->next = blockp_as_oid(dst);
+	}
+
+	if (!TOID_IS_NULL(src->next)) {
+		TX_ADD_FIELD_DIRECT(D_RW(src->next), prev);
+		D_RW(src->next)->prev = blockp_as_oid(dst);
+	}
+
+	TX_MEMCPY(dst, src, sizeof(*src));
+}
+
+static bool
+must_remove_block_a_array(struct pmemfile_vinode *vinode)
+{
+	if (vinode->first_free_block.idx != 0)
+		return false;
+
+	return vinode->first_free_block.arr !=
+	    &(D_RW(vinode->inode)->file_data.blocks);
+}
+
+static void
+remove_a_block_array(struct pmemfile_vinode *vinode)
+{
+	struct pmemfile_inode *inode;
+	TOID(struct pmemfile_block_array) to_remove;
+	TOID(struct pmemfile_block_array) new_next;
+	struct block_info *binfo;
+
+	inode = D_RW(vinode->inode);
+	binfo = &vinode->first_free_block;
+
+	to_remove = inode->file_data.blocks.next;
+
+	new_next = D_RW(to_remove)->next;
+	inode->file_data.blocks.next = new_next;
+	if (TOID_IS_NULL(new_next))
+		binfo->arr = &inode->file_data.blocks;
+	else
+		binfo->arr = D_RW(new_next);
+
+	TX_FREE(to_remove);
+	binfo->idx = binfo->arr->length;
+}
+
+void
+unlink_and_release_block_metadata(struct pmemfile_vinode *vinode,
```

Comment:
well, the allocate counter part doesn't link the prev and next fields, but this one does take care of those fields.
So, maybe unlink_and_free_block?